qiskit/algorithms/optimizers/optimizer.py

```diff
@@ -212,6 +216,81 @@ def optimize(
             )
         pass
 
+    def to_dict(self) -> Dict[str, Any]:
+        """Dump the settings of the optimizer in a dictionary for serialization."""
+        raise NotImplementedError("Dictionary conversion not implemented for this optimizer.")
+
+    @classmethod
+    def from_dict(cls, dictionary: Dict[str, Any]) -> "Optimizer":
+        """Initialize the optimizer from a serializable dictionary.
+
+        Args:
+            dictionary: The dictionary containing the name and settings for the optimizer.
+
+        Returns:
+            The initialized optimizer.
+
+        Raises:
+            ValueError: If the method on the base class ``qiskit.algorithms.optimizers.Optimizer``
+                is called but the name of the optimizer is not specified.
+        """
+        # if from_dict is called on a particular optimizer class (like SPSA) we don't need the name
+        if cls != Optimizer and issubclass(cls, Optimizer):
+            name = dictionary.get("name", None)
+            if name is not None:
+                classname = cls.__name__.lower()
+                if name.lower() != classname:
+                    raise ValueError(f"Value of the key 'name' must be '{classname}'.")
+
+            # settings are the dictionary without the key "name"
+            settings = {key: value for key, value in dictionary.items() if key != "name"}
+            return cls(**settings)
+
+        # otherwise, if from_dict is called on the Optimizer base class, we need the name
+        if "name" not in dictionary:
+            raise ValueError("Required key 'name' not found in the dictionary.")
+
+        # pylint: disable=cyclic-import
+        import qiskit.algorithms.optimizers as opt
+
+        name = dictionary["name"]
+
+        optimizers = {
+            "adam": opt.ADAM,
```

Comment:
I had mentioned this to @Cryoris  but will note here. Can we not use the classname more directly as the name - since its just part of to_dict/from_dict and does not really need to be human readable/friendly, it would avoid having to have some big lookup table here that needs updating if we add/remove an optimizer. The class name can be relative to optimizer folder so it would not be a fully qualified class name allowing anything to go in the dict.

Reply:
I'm resolving this as the de-serialization will take place in the RuntimeDecoder, so we won't have a `from_dict` method here anymore, and I'll remove the `name` from the settings too, since the RuntimeEncoder can store the class name